kernel/time/src/lib.rs

```diff
@@ -100,46 +95,55 @@ impl core::ops::Sub<Instant> for Instant {
     }
 }
 
-/// A clock frequency.
+/// A clock period, measured in femtoseconds.
 #[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
-pub struct Frequency(u64);
+pub struct Period(u64);
 
-impl Frequency {
-    /// Creates a new frequency with the specified hertz.
-    pub fn new(frequency: u64) -> Self {
-        Self(frequency)
+impl Period {
+    const MAX: Self = Self(u64::MAX);
+
+    /// Creates a new period with the specified femtoseconds.
+    pub fn new(period: u64) -> Self {
+        Self(period)
     }
 }
 
-impl From<Frequency> for u64 {
-    /// Returns the frequency in hertz.
-    fn from(f: Frequency) -> Self {
+impl From<Period> for u64 {
+    /// Returns the period in femtoseconds.
+    fn from(f: Period) -> Self {
         f.0
     }
 }
 
-impl From<u64> for Frequency {
-    /// Creates a new frequency with the specified hertz.
-    fn from(f: u64) -> Self {
-        Self(f)
+impl From<Period> for u128 {
+    /// Returns the period in femtoseconds.
+    fn from(f: Period) -> Self {
+        f.0.into()
+    }
+}
+
+impl From<u64> for Period {
+    /// Creates a new period with the specified femtoseconds.
+    fn from(period: u64) -> Self {
+        Self(period)
     }
 }
 
 /// Register the clock source that can be used to sleep when interrupts are
 /// disabled.
 ///
-/// The current early sleeper will only be overwritten by `T` if `frequency` is
-/// larger than the frequency of the current early sleeper.
+/// The provided early sleeper will overwrite the current early sleeper only if
+/// `period` is smaller than that of the current early sleeper.
 ///
 /// Returns whether the early sleeper was overwritten.
-pub fn register_early_sleeper<T>(frequency: Frequency) -> bool
+pub fn register_early_sleeper<T>(period: Period) -> bool
 where
     T: EarlySleeper,
 {
-    let old_frequency = EARLY_SLEEPER_FREQUENCY.load(Ordering::SeqCst);
-    if frequency > Frequency::new(old_frequency) {
+    let old_period = EARLY_SLEEPER_PERIOD.load();
+    if period < old_period {
```

Comment:
_technically_ we ought to use compare and swap (or whatever Rust calls it.... `compare_exchange()`) to do this.
Same goes for `register_clock_source()`.

Reply:
Compare and swap won't work because it compares equality not less than/greater than. I did change the implementation to use `fetch_update` though.